### PR TITLE
[SW-1681] Change sw version to include also patch within one h2o version

### PR DESCRIPTION
**Pull request by jakubhava**
Versions will be:

Normal: 3.26.0.9-1-2.4
where 3.26.0.9 is H2O version, -1 is iteration within the single h2o version and -2.4 is spark version.

For nightlies, the format is: 3.26.0.9-1.1-2.4, where almost everything is same as ^ except the `1.1` represents the iteration and the nightly build number after the dot.

### CODE REVIEW

File: doc/src/site/sphinx/deployment/sw_azure_dbc.rst [hunk 1]
```diff
@@ -19,7 +19,7 @@ To start Sparkling Water ``H2OContext`` on Databricks Azure, the steps are:
 
 3.  Add Sparkling Water dependency
 
-    In order to create the Java library in Databricks, go to **Libraries**, select **Maven** as the library source and type the following into the coordinates field: ``sparkling-water-package_2.11:SUBST_SW_MINOR_VERSION``.
```
Comment (jakubhava) [inline doc/src/site/sphinx/deployment/sw_azure_dbc.rst]: We need to use ``SUBST_SW_VERSION``, this was documentation bug

File: doc/build.gradle [hunk 2]
```diff
@@ -25,17 +25,11 @@ task substitute() {
...
-        def swOnlyVersion = version.tokenize("-")[0]
-        def (first, second, minorVersion) = swOnlyVersion.tokenize(".")
-        def majorVersion = "${first}.${second}"
-
         new File(siteDir).eachFileRecurse(FILES) {
             if (it.name.endsWith('.html')) {
                 def contents = file(it).getText('UTF-8')
                 contents = contents
                         .replaceAll("SUBST_SW_VERSION", version)
-                        .replaceAll("SUBST_SW_MAJOR_VERSION", majorVersion)
```
Comment (jakubhava) [inline doc/build.gradle]: after fixing the documentation bug bellow this is not used anywhere

File: jenkins/sparklingWaterPipeline.groovy [hunk 3]
```diff
@@ -10,21 +10,18 @@ def getS3Path(config) {
 
 String getNightlyVersion(config) {
     def sparkMajorVersion = config.sparkMajorVersion
-    def versionLine = readFile("gradle.properties").split("\n").find() { line -> line.startsWith('version') }
-    def version = versionLine.split("=")[1]
-    if (config.uploadNightly.toBoolean()) {
```
Comment (jakubhava) [inline jenkins/sparklingWaterPipeline.groovy]: no need to have this condition, this method is called only from uploadNightly stage

**Comment by jakubhava** [comment]
Manually tested release pipeline and nightly builds

**Comment by jakubhava** [comment]
@mn-mikke also update confluence page with release steps. There is now new checkbox when releasing sw, by default selected which is used to tell the build whether this release contain upgrade of H2O or not. It internally determines whether to upgrade minor version ( 3.26.0.1-1 -> 2.26.0.2-1) or whether to upgrade the sw patch version ( 3.26.0.1-1 -> 3.26.0.1-2)

File: py/build.gradle [hunk 4]
```diff
@@ -127,7 +127,7 @@ def copyPySetup() {
         filter {
             it.replaceAll("SUBST_SPARK_MAJOR_VERSION", sparkMajorVersion)
                     .replaceAll("SUBST_SPARK_VERSION", sparkVersion)
-                    .replaceAll("SUBST_SW_VERSION", version.substring(0, version.lastIndexOf("-")))
+                    .replaceAll("SUBST_SW_VERSION", version.substring(0, version.lastIndexOf("-").replace("_")))
```
Comment (mn-mikke) [inline py/build.gradle]: Shouldn't it be the same as [here](https://github.com/h2oai/sparkling-water/pull/1585/files#diff-bcffd8069a181d6a460a3178647d82d7R169)? or can you give me more details why it's like that?

File: doc/src/site/sphinx/migration_guide.rst [hunk 5]
```diff
@@ -80,4 +80,15 @@ Removal of Deprecated Methods and Classes
...
   ``getAlphaValue``, ``getLambda`` in favor of ``getLambdaValue``, ``setAlpha`` in favor of ``setAlphaValue`` and
-  ``setLambda`` in favor of ``setLambdaValue``. These changes ensure the consistency across Python and Scala APIs.
\ No newline at end of file
+  ``setLambda`` in favor of ``setLambdaValue``. These changes ensure the consistency across Python and Scala APIs.
+
+Change of Versioning Scheme
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Version of Sparkling Water is changed to the following pattern: ``H2OVersion-SWPatchVersion-SparkVersion``, where:
+``H2OVersion`` is full H2O Version which is integrated with this Sparkling Water. ``SWPatchVersion`` is used to specify
```
[the remark this answers is not in the document]
Comment (jakubhava) [reply doc/src/site/sphinx/migration_guide.rst]: 👍

File: doc/src/site/sphinx/migration_guide.rst [hunk 6]
```diff
@@ -80,4 +80,15 @@ Removal of Deprecated Methods and Classes
...
+patch version and ``SparkVersion`` is Spark version. This change of scheme allows us to do releases of Sparkling Water
```
[the remark this answers is not in the document]
Comment (jakubhava) [reply doc/src/site/sphinx/migration_guide.rst]: 👍

File: doc/src/site/sphinx/migration_guide.rst [hunk 7]
```diff
@@ -80,4 +80,15 @@ Removal of Deprecated Methods and Classes
...
+without the need of releasing H2O if there is only change on Sparkling Water side. In that case we just increment the
```
[the remark this answers is not in the document]
Comment (jakubhava) [reply doc/src/site/sphinx/migration_guide.rst]: 👍

File: jenkins/sparklingWaterPipeline.groovy [hunk 8]
```diff
@@ -10,21 +10,28 @@ def getS3Path(config) {
...
-        def buildNumber
-        try {
-            def lastVersion = "https://h2o-release.s3.amazonaws.com/sparkling-water/spark-${config.sparkMajorVersion}/${getS3Path(config)}latest".toURL().getText().toString()
-            def splits = lastVersion.split("-")
-            buildNumber = splits[1].toInteger() + 1
-        } catch (Exception ignored) {
-            buildNumber = 1
+    def version = readFile("gradle.properties").split("\n").find() { line -> line.startsWith("version") }.split("=")[1]
+    def h2oPart = version.split("-")[0].toString()
```
[the remark this answers is not in the document]
Comment (jakubhava) [reply jenkins/sparklingWaterPipeline.groovy]: Yup, thanks!

File: jenkins/sparklingWaterPipeline.groovy [hunk 9]
```diff
@@ -10,21 +10,28 @@ def getS3Path(config) {
...
+    def swPatch = version.split("-")[1].toString()
+    def swNightlyBuildNumber
+    try {
+        def lastVersion = "https://h2o-release.s3.amazonaws.com/sparkling-water/spark-${config.sparkMajorVersion}/${getS3Path(config)}latest".toURL().getText().toString()
+        def lastH2OPart = lastVersion.split("-")[0].toString()
```
[the remark this answers is not in the document]
Comment (jakubhava) [reply jenkins/sparklingWaterPipeline.groovy]: I don't think so, thank you!

File: jenkins/sparklingWaterPipeline.groovy [hunk 10]
```diff
@@ -10,21 +10,28 @@ def getS3Path(config) {
...
+        def lastSWPart = lastVersion.split("-")[1]
+        if (lastSWPart.contains(".")) {
+            def lastSWPatch = lastSWPart.split("\\.")[0].toString()
+            if (lastH2OPart != h2oPart || lastSWPatch != swPatch) {
+                swNightlyBuildNumber = 1 // reset the nightly build number
+            } else {
+                swNightlyBuildNumber = lastSWPatch.toInteger() + 1
```
[the remark this answers is not in the document]
Comment (jakubhava) [reply jenkins/sparklingWaterPipeline.groovy]: Really good catch, it really should!

File: py/build.gradle (hunk 4 above)
Comment (jakubhava) [reply py/build.gradle]: Yup, should be the same

**Comment by jakubhava** [comment]
Thank you @mn-mikke for really good catches, I applied fixes based on your review

**Comment by jakubhava** [comment]
@mn-mikke please have a look if all looks good

File: jenkins/sparklingWaterPipeline.groovy [hunk 11]
```diff
@@ -10,21 +10,32 @@ def getS3Path(config) {
...
-            def splits = lastVersion.split("-")
-            buildNumber = splits[1].toInteger() + 1
-        } catch (Exception ignored) {
-            buildNumber = 1
+    def version = readFile("gradle.properties").split("\n").find() { line -> line.startsWith("version") }.split("=")[1]
+    def versionParts = version.split("-")
+    def h2oPart = versionParts[0]
+    def swPatch = versionParts[1]
+    def swNightlyBuildNumber
+    try {
```
Comment (mn-mikke) [inline jenkins/sparklingWaterPipeline.groovy]: Is the following piece of code going to work with the old scheme once you merge it master?

Comment (jakubhava) [reply jenkins/sparklingWaterPipeline.groovy]: Master will be using the new scheme after we merge this. rel-3.26 will be the last one to use the old scheme

**Pull request merged**